lms/services/canvas_api.py

```diff
@@ -91,10 +94,25 @@ def public_url(self, file_id):
 
         :rtype: str
         """
-        return self._helper.validated_response(
-            self._helper.public_url_request(self._oauth2_token.access_token, file_id),
+        oauth2_token = self._oauth2_token
+
+        return self.send_with_refresh_and_retry(
+            self._helper.public_url_request(oauth2_token.access_token, file_id),
             CanvasPublicURLResponseSchema,
-        ).parsed_params["public_url"]
+            oauth2_token.refresh_token,
+        )["public_url"]
+
+    def send_with_refresh_and_retry(self, request, schema, refresh_token):
+        try:
+            return self._helper.validated_response(request, schema).parsed_params
+        except CanvasAPIAccessTokenError:
+            if not refresh_token:
+                raise
+
+            new_access_token = self.get_refreshed_token(refresh_token)
+            return self._helper.validated_response(
+                request, schema, new_access_token
+            ).parsed_params
```

Comment:
So the logic is:

1. If the request succeeds the first time then send back the response
2. If the request fails with a non-access token error (some `CanvasAPIError` that is not a `CanvasAPIAccessTokenError` then _don't_ refresh, send back an error response
3. If the request fails for access token-related reasons then attempt a refresh request
4. If the refresh request fails for any reason then send back an error response with the details of the failed refresh request on the `error_message` and `details`
5. If the refresh request succeeds then re-try the original request but with the new access token, and either send back a successful response or an error response with the `error_message` and `details` based on the second failed request

Only one refresh and retry is ever attempted per proxy API call.